File: src/lib/calibration.js
let Choices = {
  NONE: 'no calibration',
  EXISTING: 'use existing calibration',
  CALIBRATE: 'calibrate'
}

function createCalUnitMap() {
  let i = 0;
  return (vnaPort) => {
    i++;
    return [vnaPort, i];
  };
}

function createSteps(ports, calUnitSize) {
  if (calUnitSize <= 1) {
    return [];
  }
  if (ports.length == 0) {
    return [];
  }
  if (ports.length == 1) {
    return [[[ports[0],1]]];
  }

  ports.sort();
  const [commonPort] = ports.splice(0,1);
  let steps = [];
  while (ports.length) {
    let vnaPorts = null;
    if (ports.length > calUnitSize-1) {
      vnaPorts = [commonPort, ...ports.splice(0, calUnitSize-1)];
    }
    else {
      vnaPorts = [commonPort, ...ports.splice(0)];
    }
    steps.push(vnaPorts.map(createCalUnitMap()));
  }
  return steps;
}

module.exports = {
  Choices: Choices,
  createSteps: createSteps
};

File: python/lib/calibration.py
import pdb

def make_step(vnaPorts):
    result = []
    for i in range(0, len(vnaPorts)):
        result.append([vnaPorts[i], i+1])
    return result

def create_steps(ports, cal_unit_ports):
    if cal_unit_ports <= 1:
        return []
    if not ports:
        return []

    ports = ports.copy()

    if len(ports) == 1:
        return [[[*ports, 1]]]

    ports.sort()
    common_port = ports.pop(0)
    steps = []
    while ports:
        if len(ports) > cal_unit_ports - 1:
            vnaPorts = ports[0 : cal_unit_ports-1]
            vnaPorts.insert(0, common_port)
            steps.append(make_step(vnaPorts))
            ports = ports[cal_unit_ports-1:]
        else:
            vnaPorts = ports.copy()
            vnaPorts.insert(0, common_port)
            steps.append(make_step(vnaPorts))
            ports.clear()
    return steps
